### DISCUSSION TITLE
How to save waveform data onto PC as a csv file?

### DISCUSSION THREAD

**Discussion by paniash** [microsoft/Qcodes#5638] [Q&A] [answered]
Hi,

I'm trying to write a driver for the Keysight DSOX4104A oscilloscope and was wondering if there is a way to save the displayed waveform data to a csv file, not locally but to the PC controlling the instrument.

Under the [programmer's manual](https://www.keysight.com/us/en/assets/9018-06976/programming-guides/9018-06976.pdf) for the oscilloscope, the section on `:SAVE` commands mentions this but only saves locally in the oscilloscope which can be saved onto a USB device.

Alternatively, is there a way to directly read out the current waveform data from the oscilloscope to plot it on the PC? 

This may or may not be specific to this particular oscilloscope but any help would be appreciated! Thanks in advance. :)

**Comment by jenshnielsen** [comment, marked as answer]
Normally I would expect the oscilloscope to have a command to enable you to read back a waveform as data to the computer. For instance the Infinium driver here https://github.com/microsoft/Qcodes/blob/main/src/qcodes/instrument_drivers/Keysight/Infiniium.py#L227 uses `:WAV:DATA?` to read back a waveform. QCoDeS is designed for you to build parameters that enables you to get the data into memory calling get on the parameter and storing that in a qcodes dataset not so much to directly fetch csv files from the ossiloscope. But when you do store it as a qcodes dataset that can be exported to a csv file using the dataset.export method

**Reply by paniash** [reply]
Ah, that makes sense. Thanks! 

Is there a way to get the waveform data from multiple channels simultaneously? Or does that also involve using a qcodes dataset to store data from each channel.

**Reply by paniash** [reply]
Also, using the `:WAV:DATA?` command outputs a string of comma separated values. Is there a way to get it along with an array for the time axis (x-axis)?

**Reply by jenshnielsen** [reply]
> Also, using the :WAV:DATA? command outputs a string of comma separated values. Is there a way to get it along with an array for the time axis (x-axis)?

You probably have to dig into your instrument a bit more. If you look at the Infinium driver you should see that we have a special parameter to handle this. This one calculates the x-axis from start, stop and number of points  and or step-size

> Is there a way to get the waveform data from multiple channels simultaneously? Or does that also involve using a qcodes dataset to store data from each channel.

This again depends on the instrument. I would check if the instrument has a command to do this in the manual

**Reply by paniash** [reply]
Thanks a lot for your help! I'll take a look into the documentation a bit more. (the docs can get overwhelming sometimes)
